lib/ansible/modules/packaging/os/snap.py

```diff
@@ -0,0 +1,165 @@
+              like those managed by APT. This option corresponds to the --classic argument.
+        type: bool
+        required: false
+        default: False
+
+author:
+    - Victor Carceler (vcarceler@iespuigcastellar.xeill.net)
+    - Stanislas Lange (angristan) <angristan@pm.me>
+'''
+
+EXAMPLES = '''
+# Install "foo" snap
+- name: Install foo
+  snap:
+    name: foo
+
+# Remove "foo" snap
+- name: Remove VLC
+  snap:
+    name: foo
+    state: absent
+
+# Install a snap with classic confinement
+- name: Install "foo" with option --classic
+  snap:
+    name: foo
+    classic: True
+'''
+
+RETURN = '''
+msg:
+    description: An informative message on the task's output
+    type: str
+    returned: always
+    sample: "The snap 'foo' has been installed"
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+
+
+def snap_exists(module, snap_name):
+    snap_path = module.get_bin_path("snap", True)
+    cmd = "%s info %s" % (snap_path, snap_name)
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    return rc, out, err
+
+
+def is_snap_installed(module, snap_name):
+    snap_path = module.get_bin_path("snap", True)
+    cmd = "%s list %s" % (snap_path, snap_name)
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    return rc, out, err
+
+
+def install_snap(module, snap_name):
+    rc, out, err = is_snap_installed(module, snap_name)
+    if rc == 0:
+        module.exit_json(msg="The snap '%s' is already installed" % str(snap_name))
+
+    if module.params['classic']:
+        classic = '--classic'
+    else:
+        classic = ''
+
+    snap_path = module.get_bin_path("snap", True)
+    cmd = "%s install %s %s" % (snap_path, snap_name, classic)
+
+    # Actually install the snap
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    if rc == 0:
+        # snap has been installed
+        module.exit_json(msg="The snap '%s' has been installed" % str(snap_name), changed=True)
+    else:
+        # Something went wrong
+        module.fail_json(msg=err)
+
+
+def remove_snap(module, snap_name):
+    rc, out, err = is_snap_installed(module, snap_name)
+    if rc != 0:
+        module.exit_json(msg="The snap '%s' is not installed" % str(snap_name))
+
+    snap_path = module.get_bin_path("snap", True)
+    cmd = "%s remove %s" % (snap_path, snap_name)
+
+    # Actually remove the snap
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    if rc == 0:
+        # snap has been removed
+        module.exit_json(msg="The snap '%s' has been removed" % str(snap_name), changed=True)
+    else:
+        # Something went wrong
+        module.fail_json(msg=err)
+
+
+def main():
+    module_args = dict(
+        name=dict(type='str', required=True),
+        state=dict(type='str', required=False, default='present', choices=['absent', 'present']),
+        classic=dict(type='bool', required=False, default=False)
+    )
+    module = AnsibleModule(argument_spec=module_args)
```

Comment:
Should I do something besides adding `supports_check_mode=True`? I have trouble finding info about this in the docs.

Reply:
You'll need to check `if module.check_mode:` in all places where you mutate the state of target host (call install or uninstall) and instead of changing the something return module execution result as if it'd be executed for real (flags of whether smth's changed and list of removed/added packages in result).

Example with pip: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/language/pip.py#L666-L670

If you find it annoying to bypass module argument to all functions, you might want to go for class-based module implementation. Here's how I do it: https://github.com/webknjaz/ansible-gentoo-laptop/blob/master/library/eix.py#L57-L161

Reply:
I did like your first example because I find it simpler: https://github.com/ansible/ansible/pull/44939/commits/78d803f0fcddda20e7bb93cf86db21f5b2eb9c3f